server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1511,302 @@ public void testSnapshotFromStreamInput() throws IOException {
         assertEquals(ops, readOperations);
     }
 
+    public void testSnapshotCurrentHasUnexpectedOperationsForTrimmedOperations() throws Exception {
+        int extraDocs = randomIntBetween(10, 15);
+
+        // increment primaryTerm to avoid potential negative numbers
+        primaryTerm.addAndGet(extraDocs);
+        translog.rollGeneration();
+
+        long pt = primaryTerm.get();
+
+        for (int op = 0; op < extraDocs; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op, pt - op, ascii.getBytes("UTF-8"));
+            translog.add(operation);
+        }
+        try {
+            translog.trimOperations(pt, 0);
+            fail();
+        } catch (AssertionError e) {
+            assertThat(e.getMessage(),
+                is("current should not have any operations with seq#:primaryTerm [1:" + (pt - 1) + "] > 0:" + pt));
+        }
+
+        primaryTerm.incrementAndGet();
+        translog.rollGeneration();
+
+        // add a single operation to current with seq# > trimmed seq# but higher primary term
+        Translog.Index operation = new Translog.Index("test", "" + 1, 1L, primaryTerm.get(),
+            randomAlphaOfLengthBetween(1, 50).getBytes("UTF-8"));
+        translog.add(operation);
+
+        // it is possible to trim after generation rollover
+        translog.trimOperations(pt, 0);
+    }
+
+    public void testSnapshotTrimmedOperations() throws Exception {
+        List<Translog.Index> allOperations = new ArrayList<>();
+        List<Translog.Index> effectiveOperations = new ArrayList<>();
+
+        long translogOperations = 0;
+        final AtomicBoolean wasRollover = new AtomicBoolean(false);
+
+        for(int attempt = 0, maxAttempts = randomIntBetween(3, 10); attempt < maxAttempts; attempt++) {
+            int extraDocs = randomIntBetween(10, 15);
+
+            List<Translog.Index> newOperations = new ArrayList<>();
+
+            List<Long> ops = LongStream.range(0, translogOperations + extraDocs)
+                .boxed().collect(Collectors.toList());
+            Randomness.shuffle(ops);
+
+            for (final long op : ops) {
+                // use ongoing primaryTerms - or the same as it was
+                final long pt = this.primaryTerm.get();
+
+                String source = randomAlphaOfLengthBetween(1, 50);
+
+                // if there was no roll generation - have to use exactly the same source
+                // for same seq# regardless primaryTerm
+                if (!wasRollover.get()) {
```

Comment:
Instead of worrying about a roll over, I think we should look at the primary term of the existing op. If it is the same as the current term, use the same source, other wise change. This enforces our semantics regardless of the internals. Also, can you streamify this?